core/src/main/java/tech/tablesaw/io/ReadOptions.java

```diff
@@ -327,8 +357,112 @@ public Builder minimizeColumnSizes() {
+
+      return this;
+    }
+
     public ReadOptions build() {
       return new ReadOptions(this);
     }
   }
+
+  /**
+   * Allow to customize read column types. It can work in three ways:
+   *
+   * <ul>
+   *   <li>If no information is provided column types are autodetected
+   *   <li>A complete list of columns can be provided using {@link
+   *       ReadOptions.Builder#columnTypes(ColumnType[])} or {@link
+   *       ReadOptions.Builder#completeColumnTypeFunction} and they are used preventing autodetect
+   *   <li>Provide values for some column names using {@link
+   *       ReadOptions.Builder#columnType(String,ColumnType)} or {@link
+   *       ReadOptions.Builder#columnTypeByNameFunction(Function)} (String, ColumnType)}. In this
+   *       case provided columnTypes are used and the others are autodetected
+   * </ul>
+   */
+  public static class ColumnTypeReadOptions {
+    final ColumnType[] columnTypesByIdx;
+    final Map<String, ColumnType> columnTypesByNameMap;
+    final Function<String, Optional<ColumnType>> columnTypesByNameFunction;
+    final Function<String, ColumnType> completeColumnTypesByNameFunction;
+
+    public static ColumnTypeReadOptions of(ColumnType[] allColumnTypes) {
+      return new ColumnTypeReadOptions(allColumnTypes, null, null, null);
+    }
+
+    ColumnTypeReadOptions(
+        ColumnType[] columnTypesByIdx,
+        Map<String, ColumnType> columnTypesByNameMap,
+        Function<String, ColumnType> completeColumnTypesByNameFunction,
+        Function<String, Optional<ColumnType>> columnTypesByNameFunction) {
+      this.columnTypesByIdx = columnTypesByIdx;
+      this.columnTypesByNameMap = columnTypesByNameMap;
+      this.columnTypesByNameFunction = columnTypesByNameFunction;
+      this.completeColumnTypesByNameFunction = completeColumnTypesByNameFunction;
+    }
+
+    public Optional<ColumnType> columnType(int columnNumber, String columnName) {
+      Optional<ColumnType> columnType = Optional.empty();
+      if (columnTypesByIdx != null && columnNumber < columnTypesByIdx.length)
+        columnType = Optional.ofNullable(columnTypesByIdx[columnNumber]);
+      if (!columnType.isPresent() && columnTypesByNameMap != null)
+        columnType = Optional.ofNullable(columnTypesByNameMap.get(columnName));
+      if (!columnType.isPresent() && completeColumnTypesByNameFunction != null)
+        columnType = Optional.of(completeColumnTypesByNameFunction.apply(columnName));
+      if (!columnType.isPresent() && columnTypesByNameFunction != null)
+        columnType = columnTypesByNameFunction.apply(columnName);
+      return columnType;
+    }
+
+    public ColumnType[] columnTypes() {
+      return columnTypesByIdx;
+    }
+
+    public boolean canCalculateColumnTypeForAllColumns() {
+      return hasColumnTypeForAllColumns() || completeColumnTypesByNameFunction != null;
+    }
+
+    public boolean hasColumnTypeForAllColumns() {
+      return columnTypesByIdx != null && columnTypesByIdx.length > 0;
```

Comment:
this doesn't actually do what the name implies it does. If the array has one value, it would return true regardless of the number of columns.